### PR TITLE
Remove the name 'Rubinius'

### PR DESCRIPTION
**Pull request by chrisseaton**
We don't want gems finding it through feature detection, and there's no distinction between our code and theirs now - let's mix it all in together.

Basically, `Rubinius` is now `Truffle::Support`, with the idea that things can move into `Truffle` in the future (not always simple).

### CODE REVIEW

**Comment by eregon** [comment]
Please let me review before merging.
Might be good to simplify some things so we don't have deeply-nested namespaces which is verbose and cumbersome.

**Comment by chrisseaton** [comment]
@eregon if you want I can try moving most of the modules and classes into just `Truffle`. A couple of them conflict which is why I didn't do it directly here.

File: doc/user/interop.md [hunk 1]
```diff
@@ -48,14 +48,14 @@ Call `size` on the object.
 
 ### `IS_BOXED`
 
-Returns true only for instances of `String`, `Symbol`, `Rubinius::FFI::Pointer`
-and objects that respond to `unbox`.
+Returns true only for instances of `String`, `Symbol`,
+`Truffle::Support::FFI::Pointer` and objects that respond to `unbox`.
```
Comment (eregon) [inline doc/user/interop.md]: Can we simplify Truffle::Support::FFI to just Truffle::FFI? It's way too long.

File: lib/patches/bundler/bundler/ruby_version.rb [hunk 2]
```diff
@@ -15,7 +15,7 @@ def self.system
                             when "ruby"
                               ruby_version
                             when "rbx"
-                              Rubinius::VERSION.dup
+                              Truffle::Support::VERSION.dup
```
Comment (eregon) [inline lib/patches/bundler/bundler/ruby_version.rb]: This should not be modified (it's upstream code)

File: lib/rubysl/socket/rubysl.rb [hunk 3]
```diff
@@ -27,11 +27,11 @@
 module RubySL
   module Socket
     def self.bsd_support?
-      Rubinius.bsd? || Rubinius.darwin?
+      Truffle::Support.bsd? || Truffle::Support.darwin?
```
Comment (eregon) [inline lib/rubysl/socket/rubysl.rb]: I think this would be nice as `Truffle::Platform.bsd?`

File: lib/rubysl/socket/socket.rb [hunk 4]
```diff
@@ -261,7 +261,7 @@ class << self
 
   if RubySL::Socket.unix_socket_support?
     def self.pack_sockaddr_un(file)
-      max_path_size =  Rubinius::Config['rbx.platform.sockaddr_un.sun_path.size'] - 1
+      max_path_size =  Truffle::Support::Config['rbx.platform.sockaddr_un.sun_path.size'] - 1
```
Comment (eregon) [inline lib/rubysl/socket/socket.rb]: Can it be just `Truffle::Config` or `Truffle::NativeConfig` or `Truffle::Platform`? (I prefer `Truffle::Platform`)

File: src/main/java/org/truffleruby/core/string/StringNodes.java [hunk 5]
```diff
@@ -517,8 +517,8 @@ public Object sliceObjectRange(VirtualFrame frame, DynamicObject string, Dynamic
                 @Cached("new()") SnippetNode snippetNode1,
                 @Cached("new()") SnippetNode snippetNode2) {
             // TODO (nirvdrum 31-Mar-15) The begin and end values may return Fixnums beyond int boundaries and we should handle that -- Bignums are always errors.
-            final int coercedBegin = (int) snippetNode1.execute(frame, "Rubinius::Type.rb_num2int(v)", "v", Layouts.OBJECT_RANGE.getBegin(range));
-            final int coercedEnd = (int) snippetNode2.execute(frame, "Rubinius::Type.rb_num2int(v)", "v", Layouts.OBJECT_RANGE.getEnd(range));
+            final int coercedBegin = (int) snippetNode1.execute(frame, "Truffle::Support::Type.rb_num2int(v)", "v", Layouts.OBJECT_RANGE.getBegin(range));
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/StringNodes.java]: Can it be just `Truffle::Type`/

**Comment by eregon** [comment]
> A couple of them conflict which is why I didn't do it directly here.

Which ones conflict?
I think it would be best to avoid an extra namespace where possible `Truffle::Support::Type.object_class(object)` starts to look as verbose as Java code.
I think it should be done in this PR, so the diff is more straightforward and easier to review.

**Comment by chrisseaton** [comment]
For example, in `type.rb`, `Encoding` would then refer to `Truffle::Encoding` rather than `::Encoding`. It's possible that this is the only issue and I can move the rest into just `Truffle`, but that's why I didn't immediately do it.

**Comment by chrisseaton** [comment]
Shall I move the pain ones (`Type`, `FFI`, `Mirror` etc) into `Truffle` and stop when I find ones I can't move?

File: doc/user/interop.md (hunk 1 above)
Comment (chrisseaton) [reply doc/user/interop.md]: Yes, as I've said we can move things into `Truffle`, I just haven't done it yet.

File: lib/rubysl/socket/rubysl.rb (hunk 3 above)
Comment (chrisseaton) [reply lib/rubysl/socket/rubysl.rb]: Yes, again, we can move things into `Truffle` I just haven't done it yet.

File: lib/rubysl/socket/socket.rb (hunk 4 above)
Comment (chrisseaton) [reply lib/rubysl/socket/socket.rb]: You're asking the same question for all of these...

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 5 above)
Comment (chrisseaton) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: You're asking the same question for all of these...

File: lib/patches/bundler/bundler/ruby_version.rb (hunk 2 above)
Comment (chrisseaton) [reply lib/patches/bundler/bundler/ruby_version.rb]: Fixed.

**Comment by eregon** [comment]
I would rename Truffle::Encoding to Truffle::EncodingOperations like you did for Truffle::Fixnum. Yeah, I think having the non problematic ones directly under Truffle would make code nicer. Then we can think about the problematic ones and maybe simplify their namespace later as I think it's only a couple or so.

File: src/main/java/org/truffleruby/core/string/StringNodes.java (hunk 5 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/string/StringNodes.java]: Yes, I just wanted to give examples of the important ones we should simplify

File: src/main/java/org/truffleruby/language/RubyGuards.java [hunk 6]
```diff
@@ -197,7 +197,7 @@ public static boolean isRubyPointer(DynamicObject value) {
         return Layouts.POINTER.isPointer(value);
     }
 
-    public static boolean isRubiniusByteArray(DynamicObject value) {
+    public static boolean isSupportByteArray(DynamicObject value) {
```
Comment (nirvdrum) [inline src/main/java/org/truffleruby/language/RubyGuards.java]: As part of the rename, it'd be nice if this were renamed to `isTruffleByteArray`. Just something to note when you change the namespace again.

File: src/main/java/org/jcodings/transcode/TranscodingManager.java [hunk 7]
```diff
@@ -140,11 +140,11 @@ private static EConv create(byte[] sourceEncodingName, byte[] destinationEncodin
     }
 
     /**
-     * Rubinius and JRuby process Encoding::Converter options flags differently.  Rubinius splits the processing
+     * We and JRuby process Encoding::Converter options flags differently.  We split the processing
      * between initial setup and the replacement value setup, whereas JRuby handles them all during initial setup.
      * We figure out what flags JRuby additionally expects to be set and set them to satisfy EConv.
      */
-    private static int rubiniusToJRubyFlags(int flags) {
+    private static int toJRubyFlags(int flags) {
```
Comment (nirvdrum) [inline src/main/java/org/jcodings/transcode/TranscodingManager.java]: Unrelated to your change here, but we could call this `toJCodingsFlags`.

File: lib/patches/bundler/bundler/ruby_version.rb [hunk 8]
```diff
@@ -15,7 +15,7 @@ def self.system
                             when "ruby"
                               ruby_version
                             when "rbx"
-                              Truffle::Support::VERSION.dup
+                              Truffle::VERSION.dup
```
Comment (nirvdrum) [inline lib/patches/bundler/bundler/ruby_version.rb]: To avoid confusion, you might want to just squash this into the commit that changed it and remove all history that this was ever changed incorrectly.

File: src/main/java/org/truffleruby/language/RubyGuards.java (hunk 6 above)
Comment (chrisseaton) [reply src/main/java/org/truffleruby/language/RubyGuards.java]: Will do.

File: src/main/java/org/jcodings/transcode/TranscodingManager.java (hunk 7 above)
Comment (chrisseaton) [reply src/main/java/org/jcodings/transcode/TranscodingManager.java]: Will do.

**Comment by chrisseaton** [comment]
I've removed the `Truffle::Support` module, done a lot of other cleanup that this exposed, and also moved `lib/rubysl/*` into `lib/truffle` and moved everything under the `RubySL` module into `Truffle`.

**Comment by chrisseaton** [comment]
Leaving to @eregon to review again.

File: src/main/ruby/core/file.rb [hunk 9]
```diff
@@ -312,26 +312,7 @@ def chown(owner, group)
...
-      group = -1
-    end
-
-    paths.each do |path|
-      n = POSIX.lchown Truffle::Type.coerce_to_path(path), owner, group
-      Errno.handle if n == -1
-    end
-
-    paths.size
+    raise NotImplementedError, 'lchown not implemented on this platform'
```
Comment (eregon) [inline src/main/ruby/core/file.rb]: I would rather remove the first line, since the rest of the logic should work just fine.

File: lib/truffle/date.rb [hunk 10]
```diff
@@ -277,11 +277,11 @@ def <=> (other)
       when Infinity; return d <=> other.d
       when Numeric; return d
       else
-	begin
-	  l, r = other.coerce(self)
-	  return l <=> r
-	rescue NoMethodError
-	end
+	       begin
```
Comment (eregon) [inline lib/truffle/date.rb]: This looks wrong. Are there tabs in the file?

Comment (chrisseaton) [reply lib/truffle/date.rb]: There are.

File: src/main/ruby/core/file.rb (hunk 9 above)
Comment (chrisseaton) [reply src/main/ruby/core/file.rb]: It's not tested - it's dead code. But ok I'll look at running the super user tests in Docker.

**Comment by chrisseaton** [comment]
I can squash all the commits that replace `Rubinius` with `Truffle` as a first step, but after that I'm still in favour of merging the Rubinius code into ours and bring it into line with our style configuration, but will discuss that offline.

**Comment by chrisseaton** [comment]
It was more difficult than I thought to keep more commits, as they were fixes to things that only broken when I checked more tests, so I've squashed more of them.

File: lib/patches/bundler/bundler/ruby_version.rb (hunk 8 above)
Comment (eregon) [reply lib/patches/bundler/bundler/ruby_version.rb]: @chrisseaton This code should not be changed.
Probably a rebase error.

File: tool/generate-native-config.rb [hunk 11]
```diff
@@ -35,7 +35,7 @@
 # From https://github.com/rubinius/rubinius/blob/v2.71828182/rakelib/platform.rake
 # and its required files, but greatly simplified.
 
-PREFIX = 'rbx.platform'
+PREFIX = 'platform'
```
Comment (eregon) [inline tool/generate-native-config.rb]: I think we can even simplify to an empty `PREFIX`, but should be another PR.
